File: backend/app/api/documents.py
import uuid

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.schemas.document import DocumentListItem, DocumentUploadResponse
from app.services.ingestion_service import IngestionError, IngestionService


router = APIRouter(prefix="/documents", tags=["documents"])


@router.post("/upload", response_model=DocumentUploadResponse, status_code=status.HTTP_201_CREATED)
async def upload_document(
    file: UploadFile = File(...),
    source_type: str | None = Form(default=None),
    category: str | None = Form(default=None),
    title: str | None = Form(default=None),
    description: str | None = Form(default=None),
    db: Session = Depends(get_db),
) -> DocumentUploadResponse:
    service = IngestionService(db)

    try:
        return await service.index_upload(
            upload=file,
            source_type=source_type,
            category=category,
            title=title,
            description=description,
        )
    except IngestionError as exc:
        raise HTTPException(status_code=exc.status_code, detail=exc.message) from exc


@router.get("", response_model=list[DocumentListItem])
def list_documents(db: Session = Depends(get_db)) -> list[DocumentListItem]:
    service = IngestionService(db)
    return service.list_documents()


@router.delete("/{document_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_document(document_id: uuid.UUID, db: Session = Depends(get_db)) -> None:
    service = IngestionService(db)

    try:
        service.delete_document(document_id)
    except IngestionError as exc:
        raise HTTPException(status_code=exc.status_code, detail=exc.message) from exc

File: frontend/src/api/documents.ts
import { API_BASE_URL, apiGet } from "./client";
import type { DocumentListItem, DocumentUploadResponse } from "../types/document";

export type UploadDocumentInput = {
  file: File;
  sourceType?: string;
  category?: string;
  title?: string;
  description?: string;
};

export function getDocuments(signal?: AbortSignal) {
  return apiGet<DocumentListItem[]>("/documents", signal);
}

export async function uploadDocument(input: UploadDocumentInput): Promise<DocumentUploadResponse> {
  const formData = new FormData();
  formData.append("file", input.file);
  appendOptional(formData, "source_type", input.sourceType);
  appendOptional(formData, "category", input.category);
  appendOptional(formData, "title", input.title);
  appendOptional(formData, "description", input.description);

  const response = await fetch(`${API_BASE_URL}/documents/upload`, {
    method: "POST",
    body: formData,
  });

  if (!response.ok) {
    throw new Error(await getErrorMessage(response));
  }

  return response.json() as Promise<DocumentUploadResponse>;
}

export async function deleteDocument(documentId: string): Promise<void> {
  const response = await fetch(`${API_BASE_URL}/documents/${documentId}`, {
    method: "DELETE",
  });

  if (!response.ok) {
    throw new Error(await getErrorMessage(response));
  }
}

function appendOptional(formData: FormData, key: string, value?: string) {
  if (value?.trim()) {
    formData.append(key, value.trim());
  }
}

async function getErrorMessage(response: Response) {
  try {
    const payload = (await response.json()) as { detail?: string };
    return payload.detail ?? `Request failed with status ${response.status}`;
  } catch {
    return `Request failed with status ${response.status}`;
  }
}
